File: back_end/app/schemas/service_report.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field, field_validator


class UserOptionResponse(BaseModel):
    id: str
    display_name: str
    role: str


class UserOptionsListResponse(BaseModel):
    items: list[UserOptionResponse]


class ServiceReportConfigCreateRequest(BaseModel):
    name: str = Field(min_length=1, max_length=120)
    project_name: str = Field(min_length=1, max_length=120)
    report_type: str = Field(pattern="^(monthly|quarterly|annual)$")
    period_rule: str = Field(pattern="^(natural_month|natural_quarter|natural_year|custom)$")
    template_key: str = Field(
        pattern="^(ops_service_monthly_v1|ops_service_quarterly_v1|ops_service_annual_v1)$"
    )
    project_owner_user_id: str = Field(min_length=1, max_length=36)
    template_owner_user_id: str = Field(min_length=1, max_length=36)
    metric_owner_user_id: str = Field(min_length=1, max_length=36)
    enabled: bool = Field(default=True)
    recipient_emails: list[str] = Field(min_length=1)
    source_bindings: list[dict] = Field(min_length=4)

    @field_validator("source_bindings")
    @classmethod
    def validate_source_bindings(cls, value: list[dict]) -> list[dict]:
        required = {"inspection", "vulnerability", "worklog", "zentao_bug"}
        actual = {item.get("source_type") for item in value}
        if required - actual:
            raise ValueError("source_bindings must include inspection/vulnerability/worklog/zentao_bug")
        return value


class ServiceReportConfigResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    config_id: str
    name: str
    project_name: str
    report_type: str
    period_rule: str
    template_key: str
    project_owner_user_id: str
    template_owner_user_id: str
    metric_owner_user_id: str
    enabled: bool
    recipient_emails: list[str]
    source_bindings: list[dict]
    created_at: datetime
    updated_at: datetime


class ServiceReportConfigListResponse(BaseModel):
    items: list[ServiceReportConfigResponse]
    page: int
    page_size: int
    total: int


class SourceResultResponse(BaseModel):
    source_type: str
    status: str
    record_count: int = 0
    valid_row_count: int = 0
    invalid_row_count: int = 0
    error_message: str | None = None


class ServiceReportSourceRunResponse(BaseModel):
    source_run_id: str
    config_id: str
    status: str
    window_start: datetime
    window_end: datetime
    included_sources: list[str]
    source_results: list[SourceResultResponse]
    created_at: datetime


class ServiceReportSourceRunCreateForm(BaseModel):
    window_start: datetime
    window_end: datetime


class SectionResponse(BaseModel):
    key: str
    title: str
    data_status: str
    blocking_reason: str | None = None
    content_markdown: str


class ExportArtifactResponse(BaseModel):
    format: str
    file_name: str
    download_url: str
    generated_at: datetime


class ServiceReportRunCreateRequest(BaseModel):
    window_start: datetime
    window_end: datetime
    source_run_id: str
    force_regenerate: bool = False


class ServiceReportRunCreateResponse(BaseModel):
    run_id: str
    config_id: str
    source_run_id: str
    status: str
    completeness_status: str
    window_start: datetime
    window_end: datetime
    report_type: str
    template_key: str
    created_at: datetime


class ServiceReportRunListItemResponse(BaseModel):
    run_id: str
    config_id: str
    config_name: str
    project_name: str
    report_type: str
    status: str
    completeness_status: str
    window_start: datetime
    window_end: datetime
    source_run_id: str
    export_formats: list[str]
    created_at: datetime
    finished_at: datetime | None = None


class ServiceReportRunListResponse(BaseModel):
    items: list[ServiceReportRunListItemResponse]
    total: int
    page: int
    page_size: int


class ServiceReportRunDetailResponse(BaseModel):
    run_id: str
    config_id: str
    config_snapshot: dict
    source_run_id: str
    status: str
    completeness_status: str
    window_start: datetime
    window_end: datetime
    source_snapshot_summary: dict
    report_payload: dict
    manual_note: str | None = None
    export_artifacts: list[ExportArtifactResponse]
    evidence_refs: list[dict]
    error_message: str | None = None
    created_at: datetime
    finished_at: datetime | None = None


class ManualNoteUpdateRequest(BaseModel):
    manual_note: str | None = Field(default=None, max_length=2000)


class ManualNoteUpdateResponse(BaseModel):
    run_id: str
    manual_note: str | None = None


class ExportRequest(BaseModel):
    format: str = Field(pattern="^(markdown|html)$")
    overwrite: bool = False


class ExportResponse(BaseModel):
    run_id: str
    format: str
    file_name: str
    download_url: str
    generated_at: datetime

File: front_end/src/api/serviceReport.ts
import axios from 'axios';
import request from './request';
import type {
  ServiceReportConfig,
  ServiceReportExportArtifact,
  ServiceReportRunDetail,
  ServiceReportRunItem,
  ServiceReportSourceRun,
  ServiceReportType,
  UserOption,
} from './types';

export interface ServiceReportConfigCreatePayload {
  name: string;
  project_name: string;
  report_type: ServiceReportType;
  period_rule: 'natural_month' | 'natural_quarter' | 'natural_year' | 'custom';
  template_key: 'ops_service_monthly_v1' | 'ops_service_quarterly_v1' | 'ops_service_annual_v1';
  project_owner_user_id: string;
  template_owner_user_id: string;
  metric_owner_user_id: string;
  enabled: boolean;
  recipient_emails: string[];
  source_bindings: Array<{
    source_type: 'inspection' | 'vulnerability' | 'worklog' | 'zentao_bug';
    ingest_mode: 'file_import';
  }>;
}

export interface CreateServiceReportSourceRunPayload {
  window_start: string;
  window_end: string;
  inspection_file: File;
  vulnerability_file: File;
  worklog_file: File;
  zentao_bug_file: File;
}

export interface CreateServiceReportRunPayload {
  window_start: string;
  window_end: string;
  source_run_id: string;
  force_regenerate?: boolean;
}

export interface ServiceReportRunsQuery {
  config_id?: string;
  report_type?: ServiceReportType;
  status?: string;
  page?: number;
  page_size?: number;
}

export const getUserOptions = () => {
  return request.get<unknown, { items: UserOption[] }>('/users/options');
};

export const getServiceReportConfigs = (params?: {
  report_type?: ServiceReportType;
  enabled?: boolean;
  keyword?: string;
  page?: number;
  page_size?: number;
}) => {
  return request.get<unknown, { items: ServiceReportConfig[]; total: number; page: number; page_size: number }>(
    '/service-report-configs',
    { params },
  );
};

export const createServiceReportConfig = (payload: ServiceReportConfigCreatePayload) => {
  return request.post<unknown, ServiceReportConfig>('/service-report-configs', payload);
};

export const createServiceReportSourceRun = (configId: string, payload: CreateServiceReportSourceRunPayload) => {
  const formData = new FormData();
  formData.append('window_start', payload.window_start);
  formData.append('window_end', payload.window_end);
  formData.append('inspection_file', payload.inspection_file);
  formData.append('vulnerability_file', payload.vulnerability_file);
  formData.append('worklog_file', payload.worklog_file);
  formData.append('zentao_bug_file', payload.zentao_bug_file);

  return request.post<unknown, ServiceReportSourceRun>(
    `/service-report-configs/${configId}/source-runs`,
    formData,
    { headers: { 'Content-Type': 'multipart/form-data' } },
  );
};

export const createServiceReportRun = (configId: string, payload: CreateServiceReportRunPayload) => {
  return request.post<unknown, {
    run_id: string;
    config_id: string;
    source_run_id: string;
    status: string;
    completeness_status: string;
    window_start: string;
    window_end: string;
    report_type: string;
    template_key: string;
    created_at: string;
  }>(`/service-report-configs/${configId}/report-runs`, payload);
};

export const getServiceReportRuns = (params?: ServiceReportRunsQuery) => {
  return request.get<unknown, { items: ServiceReportRunItem[]; total: number; page: number; page_size: number }>(
    '/service-report-runs',
    { params },
  );
};

export const getServiceReportRunDetail = (runId: string) => {
  return request.get<unknown, ServiceReportRunDetail>(`/service-report-runs/${runId}`);
};

export const updateServiceReportManualNote = (runId: string, manual_note: string | null) => {
  return request.patch<unknown, { run_id: string; manual_note: string | null }>(
    `/service-report-runs/${runId}/manual-note`,
    { manual_note },
  );
};

export const createServiceReportExport = (runId: string, format: 'markdown' | 'html', overwrite = false) => {
  return request.post<unknown, {
    run_id: string;
    format: 'markdown' | 'html';
    file_name: string;
    download_url: string;
    generated_at: string;
  }>(`/service-report-runs/${runId}/export`, { format, overwrite });
};

export const downloadServiceReportExport = async (
  runId: string,
  artifact: Pick<ServiceReportExportArtifact, 'format' | 'file_name'>,
) => {
  const token = localStorage.getItem('token');
  const response = await axios.get(`/api/v1/service-report-runs/${runId}/export`, {
    params: { format: artifact.format },
    responseType: 'blob',
    headers: token ? { Authorization: `Bearer ${token}` } : undefined,
  });

  const blob = new Blob([response.data], {
    type: artifact.format === 'html' ? 'text/html;charset=utf-8' : 'text/markdown;charset=utf-8',
  });
  const url = window.URL.createObjectURL(blob);
  const link = document.createElement('a');
  link.href = url;
  link.download = artifact.file_name;
  document.body.appendChild(link);
  link.click();
  document.body.removeChild(link);
  window.URL.revokeObjectURL(url);
};
